s3url/s3url.go

```diff
@@ -0,0 +1,195 @@
+// Package url defines URL types and has helper methods to parse a string into URLs.
+package s3url
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+const (
+	// s3WildCharacters is valid wildcard characters for a S3Url
+	s3WildCharacters string = "?*"
+
+	// s3Prefix is the prefix used on s3 URLs
+	s3Prefix string = "s3://"
+
+	// s3Separator is the path separator for s3 URLs
+	s3Separator string = "/"
+
+	// matchAllRe is the regex to match everything
+	matchAllRe string = ".*"
+)
+
+// S3Url is the metadata that is used on S3 operations (listing, querying)
+type S3Url struct {
+	Bucket      string
+	Delimiter   string
+	Key         string
+	Prefix      string
+	filter      string
+	filterRegex *regexp.Regexp
+}
+
+func (s S3Url) String() string {
+	return "s3://" + s.Bucket + "/" + s.Key
+}
+
+// Format formats the S3Url to the format "<bucket>[/<key>]".
+func (s S3Url) Format() string {
+	if s.Key == "" {
+		return s.Bucket
+	}
+	return s.Bucket + s3Separator + s.Key
+}
+
+// setFilterRegex creates url metadata for both wildcard and non-wildcard operations.
+//
+// It converts wildcard strings to regex format
+// and pre-compiles it for later usage. It is default to
+// ".*" to match every key on S3.
+//
+// filter is the part that comes after the wildcard string.
+// prefix is the part that comes before the wildcard string.
+//
+// Example:
+//		key: a/b/test?/c/*.tsv
+//		prefix: a/b/test
+//		filter: ?/c/*
+//		regex: ^a/b/test./c/.*?\\.tsv$
+//		delimiter: ""
+//
+// It prepares delimiter, prefix and regex for regular strings.
+// These are used in S3 listing operations.
+// See: https://docs.aws.amazon.com/AmazonS3/latest/dev/ListingKeysHierarchy.html
+//
+// Example:
+//		key: a/b/c
+//		prefix: a/b/c
+//		filter: ""
+//		regex: ^a/b/c.*$
+//		delimiter: "/"
+//
+func (s *S3Url) setPrefixAndFilter() error {
+	loc := strings.IndexAny(s.Key, s3WildCharacters)
+	wildOperation := loc > -1
+	if !wildOperation {
+		s.Delimiter = s3Separator
+		s.Prefix = s.Key
+	} else {
+		s.Prefix = s.Key[:loc]
+		s.filter = s.Key[loc:]
+	}
+
+	filterRegex := matchAllRe
+	if s.filter != "" {
+		filterRegex = regexp.QuoteMeta(s.filter)
+		filterRegex = strings.Replace(filterRegex, "\\?", ".", -1)
+		filterRegex = strings.Replace(filterRegex, "\\*", ".*?", -1)
+	}
+	filterRegex = s.Prefix + filterRegex
+	r, err := regexp.Compile("^" + filterRegex + "$")
+	if err != nil {
+		return err
+	}
+	s.filterRegex = r
+	return nil
+}
+
+// Clone creates a new s3url with the values from the receiver.
+func (s S3Url) Clone() S3Url {
+	return S3Url{
+		Bucket:      s.Bucket,
+		Delimiter:   s.Delimiter,
+		Key:         s.Key,
+		Prefix:      s.Prefix,
+		filter:      s.filter,
+		filterRegex: s.filterRegex,
+	}
+}
+
+// Match check if given key matches with regex and
+// returns parsed key.
+func (s S3Url) Match(key string) (string, bool) {
```

Comment:
Can we update the signature of `Match` function like the following : 

`func (s S3Url) Match(key string) string` 

It can return empty string if there is not any match.  This will simplify the usage and test cases.

Reply:
done